### PR TITLE
feat: Enhance Helm chart with multi-deployment, SVC, HPA, and ingress

### PR DESCRIPTION
**Pull request by MahaGamal**
- **Multi-Deployment Support**: Enabled the chart to handle multiple deployments, allowing to define and manage several instances of the application easily.
- **Service Configuration**:
  - Created Kubernetes services for each deployment to expose applications and manage traffic.
- **Horizontal Pod Autoscalers (HPA)**:
  - Integrated HPA configurations to enable automatic scaling and customize scaling policies through values.yaml.
- **Service Accounts**:
  - Added the ability to create and associate service accounts with deployments for improved security and access control.
  - Ensured service accounts can be customized or disabled via values.yaml.
- **Ingress Resources**:
  - Implemented ingress configurations for managing external access to services, including TLS support.
  - Define hosts and paths for routing traffic in values.yaml.
- **Values.yaml Updates**:
  - Extended values.yaml to accommodate new configurations for deployments, services, HPA, service accounts, and ingress.
  - Provided default values for ease of use while allowing customization.

This commit enhances the Helm chart’s capabilities, making it user-friendly for deploying multi applications in Kubernetes.

### CODE REVIEW

File: values.yaml [hunk 1]
```diff
@@ -18,221 +18,223 @@ extraLabels: {}
...
-  cpuAverageUtilization: 50
-
-
-# add your custom metrics confgurations here
-  customMetrics: {}
-
-# add your custom behavior confgurations here
-  customBehavior: {}
+# Configuration for servers deployment to create one or more deployment.
+Servers:
```
Comment (ThisIsQasim) [inline values.yaml:26]: can you make it lowercase?

File: templates/servicemonitor.yaml [hunk 2]
```diff
@@ -1,28 +1,30 @@
...
+      app: {{ include "<CHARTNAME>.name" $ }}
+      release: {{ $.Release.Name }}
   namespaceSelector:
     matchNames:
-    - {{ .Release.Namespace }}
+    - {{ $.Release.Namespace }}
   endpoints:
     - port: http
       interval: 10s
       path: /prometheus
```
Comment (ThisIsQasim) [inline templates/servicemonitor.yaml]: Should we make the path and port configurable?

File: templates/servicemonitor.yaml [hunk 3]
```diff
@@ -1,28 +1,30 @@
-{{- if .Values.serviceMonitor.enabled -}}
+{{- range $name, $values := .Values.Servers }}
+{{- if $values.serviceMonitor.enabled -}}
 apiVersion: monitoring.coreos.com/v1
 kind: ServiceMonitor
 metadata:
-  name: {{ include "<CHARTNAME>.fullname" . }}-sm
-  namespace: {{ $.Values.serviceMonitor.namespace }}
+  name: {{ include "<CHARTNAME>.fullname" $ }}-{{ $name }}-sm
```
Comment (ThisIsQasim) [inline templates/servicemonitor.yaml]: `-sm` suffix is redundant
```suggestion
  name: {{ include "<CHARTNAME>.fullname" $ }}-{{ $name }}
```

File: templates/serviceaccount.yaml [hunk 4]
```diff
@@ -1,15 +1,17 @@
-{{- if .Values.serviceAccount.enabled -}}
+{{- range $name, $values := .Values.Servers }}
+{{- if $values.serviceAccount -}}
 apiVersion: v1
 kind: ServiceAccount
 metadata:
-  name: {{ include "<CHARTNAME>.fullname" . }}-sa
+  name: {{ include "<CHARTNAME>.fullname" $ }}-{{ $name }}-sa
```
Comment (ThisIsQasim) [inline templates/serviceaccount.yaml]: `-sm` suffix is redundant
```suggestion
  name: {{ include "<CHARTNAME>.fullname" $ }}-{{ $name }}
```

File: templates/service.yaml [hunk 5]
```diff
@@ -1,27 +1,29 @@
...
       {{- toYaml $value | nindent 6 }}
   {{- end }}
   selector:
-    app.kubernetes.io/name: {{ include "<CHARTNAME>.name" . }}
-    app.kubernetes.io/instance: {{ .Release.Name }}
-    app.kubernetes.io/component: server
+    app.kubernetes.io/name: {{ include "<CHARTNAME>.name" $ }}
+    app.kubernetes.io/instance: {{ $.Release.Name }}
+    app.kubernetes.io/component: {{ $name }}
+{{- end }}
```
Comment (ThisIsQasim) [inline templates/service.yaml]:
```suggestion
{{- end }}

```

File: templates/deployment.yaml [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+              valueFrom: {{ tpl $value $ }}
+            {{- end }}
+            {{- end }}
+            {{- if $.Values.global }}
+            {{- range $key, $value := $.Values.global.env }}
+            - name: {{ $key }}
+              value: {{ tpl $value $ | quote }}
+            {{- end }}
+            {{- end }}
+          envFrom:
```
Comment (ThisIsQasim) [inline templates/deployment.yaml:108]: can you add a ref to the onePassword generated secret here?

**Review by ThisIsQasim — COMMENTED** [review]
Minor suggestions but overall looks really good

File: templates/deployment.yaml (hunk 6 above)
Comment (MahaGamal) [reply templates/deployment.yaml:108]: This setup enables adding multiple secrets with a flexible, generic structure. Each entry under envFrom specifies secretRef which can be referenced in values . This configuration supports 1Password or other secret providers dynamically. like  https://github.com/mrsool/app_charts/blob/master/auth/values.yaml#L117

File: templates/service.yaml (hunk 5 above)
Comment (MahaGamal) [reply templates/service.yaml]: noted!

File: values.yaml (hunk 1 above)
Comment (MahaGamal) [reply values.yaml:26]: I recommend keeping `Servers` in mixed case as the title for all deployment names, while the actual deployment names should be in lowercase so switching to lowercase may cause confusion

File: templates/servicemonitor.yaml (hunk 2 above)
Comment (MahaGamal) [reply templates/servicemonitor.yaml]: yes, it's easy

File: values.yaml (hunk 1 above)
Comment (ThisIsQasim) [reply values.yaml:26]: the convention for values files is camelCase

File: templates/deployment.yaml (hunk 6 above)
Comment (ThisIsQasim) [reply templates/deployment.yaml:108]: I think a ref to the onePassword generated secret should be automatically added in addition to other optional secrets. i.e. if `environmentVariablesOnePasswordPath` is defined then
```
            - secretRef:
                 name: {{ include "<CHARTNAME>.fullname" $ }}-{{ $name }}-environment-variables`
```
should be added automatically

File: values.yaml [hunk 7]
```diff
@@ -148,6 +148,9 @@ Servers:
     serviceMonitor:
       enabled: false
       namespace: monitoring
+      port: htto
```
Comment (ThisIsQasim) [inline values.yaml]:
```suggestion
      port: http
```

File: templates/deployment.yaml (hunk 6 above)
Comment (MahaGamal) [reply templates/deployment.yaml:108]: Didn't get it? what you mention i already add `if environmentVariablesOnePasswordPath` condition

Comment (ThisIsQasim) [reply templates/deployment.yaml:108]: I had commented before you added that 🫤

File: templates/deployment.yaml [hunk 8]
```diff
@@ -0,0 +1,155 @@
+{{- range $name, $values := .Values.Servers }}
+{{- $fullName := include "<CHARTNAME>.fullname" $ }}  # Use $ to access the root context
+{{- $releasePrefix := include "<CHARTNAME>.releasePrefix" $ }}
+{{- $_ := set $.Release "Prefix" $releasePrefix }}
```
Comment (ThisIsQasim) [inline templates/deployment.yaml:4]: You have to add `---` if you want to generate multiple resources from a single file
```suggestion
{{- $_ := set $.Release "Prefix" $releasePrefix }}
---
```

Comment (MahaGamal) [reply templates/deployment.yaml:4]: Good catch, Yes i forget to add

File: values.yaml (hunk 1 above)
Comment (MahaGamal) [reply values.yaml:26]: could you point it to me

Comment (ThisIsQasim) [reply values.yaml:26]: https://helm.sh/docs/chart_best_practices/values/#naming-conventions

Comment (MahaGamal) [reply values.yaml:26]: This is already done with names and sub-names, Each item under this structure is like documentation however this is different case, We should format it to be easily readable with titles styled.

**Pull request merged**